### PR TITLE
Download Summary CSV Report for the given Report ID

### PR DESCRIPTION
**Pull request by kirankumbhar**
| Title                          | Value                                                                                     |
| ------------------------------ | ----------------------------------------------------------------------------------------- |
| **Type**                       | Feature                                              |
| **Ticket/Issue**               | PD-686                                      |
| **Migrations**                 | No                    |
| **Migration&nbsp;Scripts**     | No                                              |
| **ENV&nbsp;vars&nbsp;change**  | No |
| **Frontend**                   | No                                                         |
| **Local&nbsp;testing**         | Done                                                                      |
| **Staging&nbsp;testing**       | Skip                                              |
| **On&nbsp;premise&nbsp;notes** | NA                              |
| **Documentation**              | None                      |
| **Release&nbsp;notes**         | None                                   |
| **Version&nbsp;upgrade**       | Major                                                                    |

---

### Changelog

1. Added download report summary csv command
    ```bash
    appknox reports download summary-csv <reportID> --output /path/to/save/file.csv
    ```

### CODE REVIEW

File: cmd/reports.go [hunk 1]
```diff
@@ -37,7 +37,39 @@ var reportsListCmd = &cobra.Command{
...
+	Short: "Download Reports",
+	Long:  `Download reports in different formats such as CSV, excel or PDF`,
+}
+
+var reportsDownloadCsvCmd = &cobra.Command{
+	Use:   "summary-csv",
+	Short: "List Vulnerability Analysis Reports",
+	Args: func(cmd *cobra.Command, args []string) error {
+		if len(args) < 1 {
+			return errors.New("file id is required")
```
Comment (rajan262) [inline cmd/reports.go]: The error message should be `Report ID is required`

File: cmd/reports.go [hunk 2]
```diff
@@ -37,7 +37,39 @@ var reportsListCmd = &cobra.Command{
...
+		}
+		return nil
+	},
+	Run: func(cmd *cobra.Command, args []string) {
+		reportID, err := strconv.Atoi(args[0])
+		if err != nil {
+			err := errors.New("Valid file id is required")
```
Comment (rajan262) [inline cmd/reports.go]: The error message here should also be `Report ID is required`

File: cmd/reports.go [hunk 3]
```diff
@@ -37,7 +37,39 @@ var reportsListCmd = &cobra.Command{
...
+		outputFilePath, _ := cmd.Flags().GetString("output")
+		helper.ProcessDownloadReportCSV(reportID, outputFilePath)
+		if err != nil {
+			helper.PrintError(err)
+		}
+	},
+}
+
 func init() {
+	reportsDownloadCsvCmd.Flags().StringP("output", "o", "", "Output directory to save reports")
```
Comment (rajan262) [inline cmd/reports.go]: Let's keep this output flag to reportDownload command level. We'll use the same flag to excel file as well.

File: helper/reports.go [hunk 4]
```diff
@@ -55,3 +57,37 @@ func ProcessListReports(fileID int) error {
 	return nil
 
 }
+
+func ProcessDownloadReportCSV(reportID int, outputDirPath string) {
+	id := strconv.Itoa(reportID)
+	fmt.Println("Fetching Summary CSV Report for ID " + id + " ...")
```
Comment (rajan262) [inline helper/reports.go]: Check for verbose option and only do extra printing if versbose switch is on.

File: appknox/reports.go [hunk 5]
```diff
@@ -68,3 +80,60 @@ func (s *ReportsService) List(ctx context.Context, fileID int) ([]*ReportResult,
...
+		}
+		if err != nil {
+			return err
+		}
+		fmt.Println(string(line))
+	}
+}
+
+//Output report from buffer to file
+func (s *ReportsService) WriteReportDataToFile(respBody io.ReadCloser, reportID string, fileDirPath string) (string, error) {
```
Comment (rajan262) [inline appknox/reports.go]: The output file name will be provided in the `output` option. We don't have to create our own file name.

File: appknox/appknox.go [hunk 6]
```diff
@@ -244,6 +244,12 @@ func (c *Client) Do(ctx context.Context, req *http.Request, v interface{}) (*Res
 	return response, err
 }
 
+// DoTxt sends an API request and returns the API response.
+// The API response is plaintext.
+func (c *Client) DoTxt(request *http.Request) (*http.Response, error) {
```
Comment (rajan262) [inline appknox/appknox.go]: This method is doing nothing other than calling `client.Do` method. Why do we need to have this separate method?

File: appknox/reports.go [hunk 7]
```diff
@@ -68,3 +80,60 @@ func (s *ReportsService) List(ctx context.Context, fileID int) ([]*ReportResult,
...
+	}
+	return drfResponseReportDownloadUrl.Url, err
+
+}
+
+//Download Report Data from Url to buffer
+func (s *ReportsService) DownloadReportData(ctx context.Context, downloadUrl string) (*http.Response, error) {
+
+	request, err := s.client.NewRequest("GET", downloadUrl, nil)
+	resp, err := s.client.Reports.client.DoTxt(request)
```
Comment (rajan262) [inline appknox/reports.go]: Make it generic so that it can download any type of report.

Comment (kirankumbhar) [reply appknox/reports.go]: This is a generic function that will download report from the signed download URL to the response body buffer. can be used any type of report where response is report data itself irrespective of format

File: cmd/reports.go (hunk 1 above)
Comment (kirankumbhar) [reply cmd/reports.go:51]: Updated

Comment (kirankumbhar) [reply cmd/reports.go]: You list reports using file id not the report id

File: cmd/reports.go (hunk 2 above)
Comment (kirankumbhar) [reply cmd/reports.go]: This is updated 👍🏼

File: cmd/reports.go (hunk 3 above)
Comment (kirankumbhar) [reply cmd/reports.go]: Done 👍🏼

File: appknox/reports.go (hunk 5 above)
Comment (kirankumbhar) [reply appknox/reports.go]: Done 👍🏼

File: helper/reports.go (hunk 4 above)
Comment (kirankumbhar) [reply helper/reports.go]: removed all print statements for now. Will come back to verbose option later on

File: cmd/reports.go (hunk 1 above)
Comment (rajan262) [reply cmd/reports.go]: But this is not list command. It is `download summary-csv` command 😛

File: appknox/reports.go [hunk 8]
```diff
@@ -68,3 +80,60 @@ func (s *ReportsService) List(ctx context.Context, fileID int) ([]*ReportResult,
...
+		fmt.Println(string(line))
+	}
+}
+
+//Output report from buffer to file
+func (s *ReportsService) WriteReportDataToFile(respBody io.ReadCloser, outputFilePath string) (string, error) {
+
+	filePath := filepath.FromSlash(outputFilePath)
+	dirPath := filepath.Dir(filePath)
+	os.MkdirAll(dirPath, os.ModePerm)
```
Comment (rajan262) [inline appknox/reports.go]: Handle error here as well. The directory creation might fail due to permission error.

File: appknox/reports.go [hunk 9]
```diff
@@ -44,6 +49,13 @@ type DRFResponseReport struct {
 	Previous string          `json:"previous,omitempty"`
 	Results  []*ReportResult `json:"results"`
 }
+type DRFResponseReportDownloadUrl struct {
+	Url string `json:"url"`
+}
+
+type DRFResponseReportDataCSV struct {
```
Comment (rajan262) [inline appknox/reports.go]: This is not being used anywhere.

File: appknox/reports.go (hunk 8 above)
Comment (kirankumbhar) [reply appknox/reports.go]: This returns nil if not able to make directories. hence error is handled at writing time

File: helper/reports.go [hunk 10]
```diff
@@ -55,3 +57,26 @@ func ProcessListReports(fileID int) error {
...
+	if err != nil {
+		return err
+	}
+	if outputFilePath != "" {
+		_, err := client.Reports.WriteReportDataToFile(resp.Body, outputFilePath)
+		if err != nil {
+			return errors.New(fmt.Sprintf("Failed to download report. Error: %v", err))
+		}
+	}
+	err = client.Reports.WriteReportDataToTerminal(resp)
```
Comment (rajan262) [inline helper/reports.go]: I'm not sure if we should print date to the terminal in both cases. I think we should only do one thing. Either write to file or write to stdout.

File: cmd/reports.go (hunk 1 above)
Comment (kirankumbhar) [reply cmd/reports.go]: My bad. This is corrected

File: appknox/appknox.go (hunk 6 above)
Comment (kirankumbhar) [reply appknox/appknox.go]: Removed this method

File: appknox/reports.go (hunk 8 above)
Comment (kirankumbhar) [reply appknox/reports.go]: Error handling is added although the chances of error coming is very less

File: helper/reports.go (hunk 10 above)
Comment (kirankumbhar) [reply helper/reports.go]: Yes, early exit is added with no error

File: appknox/reports.go (hunk 9 above)
Comment (kirankumbhar) [reply appknox/reports.go]: Removed 👍🏼

File: appknox/reports.go (hunk 8 above)
Comment (rajan262) [reply appknox/reports.go]: Totally agree 👍

**Review by rajan262 — DISMISSED** [review]

**Review by cosmosgenius — APPROVED** [review]

**Pull request merged**